l10n_it_fatturapa/models/account.py

```diff
@@ -197,6 +197,9 @@ def create(self, vals_list):
                 vals["lineRef"] = line.sequence
         return super().create(vals_list)
 
+    def setlineRef(self, n):
+        self.lineRef = n
```

Comment:
`lineRef` dovrebbe essere assegnato correttamente, non assegnato e poi corretto al momento della generazione della fattura elettronica.
Come mai il valore assegnato poco sopra è da correggere?

Reply:
Il progressivo della riga fattura nell'XML non è lo stesso della sequence di Odoo, purtroppo. Il problema si manifesta solo nell'XML. Il valore reale di `lineRef` si conosce solo all'interno dell'XML, dipende da come l'XML loopa sulle righe.    Ripetendo la logica fuori dall'XML lo si potrebbe "indovinare", ma si crea una duplicazione del codice e della logica che poi bisogna ricordarsi di mantenere allineata dovesse cambiare nell'XML.

In effetti nella `create()` l'assegnazione della `lineRef` non serve a nulla, ma comunque il problema a monte è più generale, l'interfaccia per la gestione dei documenti collegati ad una singola riga manca completamente.

Al momento l'utente può inserire documenti collegati **alla fattura** e impostare a mano il lineRef.  A livello di XML non ci dovrebbe essere differenza.

Tuttavia a livello di nostro codice, è previsto sia il link tra fatture e documenti sia tra righe fattura e documenti.

Questa seconda parte di fatto esiste solo ad uso di altri moduli che da python creano tale relazione che ha perfettamente senso sia per riga in certi casi. Per es. una riga negativa relativa ad un acconto deve riportare il dati della fattura di acconto collegata.

Ci ho guardato, non è che l'associazione a livello di righe sia estemporanea solo in un punto:
https://github.com/OCA/l10n-italy/blob/a7ccab05455aec05ea66f699b970c85018de9522/l10n_it_fatturapa/models/account.py#L171-L184
per dire la usa anche `l10n_it_fatturapa_sale`:
https://github.com/OCA/l10n-italy/blob/a7ccab05455aec05ea66f699b970c85018de9522/l10n_it_fatturapa_sale/models/sale_order_line.py#L10-L15
dove i documenti sono collegati ad una riga d'ordine e poi vengono associati alla riga fattura:
https://github.com/OCA/l10n-italy/blob/a7ccab05455aec05ea66f699b970c85018de9522/l10n_it_fatturapa_sale/models/sale_order_line.py#L25-L35


La scelta qui è: o si fa refactoring e smontiamo completamente la parte che associa righe e documenti, lasciando solo quella tra fattura e documenti, oppure si corregge con il minimo intervento possibile il bug in questione.

Questa PR si limita a ripristinare il comportamento della 14, che è stato rotto dalla migrazione.

Del refactoring e dell'eliminazione dei documenti collegati per riga (eventualmente - non che io sia necessariamente d'accordo) se ne deve parlare altrove, IMHO.

Qui il problema è che nella 14 `lineRef` veniva assegnato del q-web (almeno nelle intenzioni) e adesso la 16 non te lo lascia fare direttamente, per cui ho risolto con la modifica minima di creare un metodo di set().

Reply:
Capito, grazie.
Penso sarebbe meglio assegnare un ordine lato codice e poi limitarsi a rappresentarlo nel XML, ma visto che questa PR mira semplicemente a ripristinare un comportamento della `14.0` non è questo il posto per farlo.

Giusto per info, per associare documenti correlati alle righe fattura è sufficiente abilitare la visualizzazione "form" delle righe fattura nel form della fattura (in pratica rimuovere `editable="bottom"` da https://github.com/odoo/odoo/blob/d0876de81dc66cb6483598deb33bf6c81212ec89/addons/account/views/account_move_views.xml#L874)

Reply:
Il discorso è il n. riga è interno all'XML. Prima, puoi solo replicare la logica e sperare che sia uguale.

In realtà, in Odoo standard hanno un approccio diverso, se non erro, passano una `list` con le righe al template (anche se credo che abbiamo la limitazione una fattura = un XML).

Visto che valutiamo di allinearci, fare un refactoring non ha molto senso in questa fase.  Avendo una `list`, di fatto basta un `enumerate` su quella e lineRef è l'indice +1 e quindi può essere determinato in anticipo con maggior certezza che poi l'XML non ci metterà del suo (e se qualche riga fattura non va messa nell'XML per qualsiasi ragione, basta non mettarla nella `list` - noi "lavoriamo" direttamente su `invoice_line_ids`, il che ha dei pro e dei contro).

Al momento la nostra logica è:
```
record.invoice_line_ids.sorted(key=lambda l: (-l.sequence, l.date, l.move_name, -l.id), reverse=True)
```
già mi secca che all'interno del template sia duplicata, presente in due punti (e vanno tenuti allineati). Creare un altro duplicato in file python mi piace ancora meno.  Andrebbe trasformata in un metodo, ma appunto meglio vedere prima che succede con l'integrazione Odoo.